content/en/blog/_posts/2023-12-05-image-filesystem.md

```diff
@@ -0,0 +1,170 @@
+This allows for more flexibility in configuring Kubernetes and provides support for adding a larger disk for the container storage while keeping the default filesystem untouched.  
+
+One area we need to explain is where/what Kubernetes is writing to disk.
+
+Kubernetes has persistent data and ephemeral data.  The root filesystem of Kubernetes is configurable, but it is usually assumed to be `/var/lib/kubelet`.  In the Kubernetes docs, this is sometimes referred to as the root or node filesystem.  The bulk of this data can be categorized into ephemeral storage, logs and container runtime.
+
+**Ephemeral storage**
+
+Pods and containers can require temporary or transient local storage for their operation. The lifetime of this ephemeral storage does not extend beyond the life of the individual pod, and this ephemeral storage cannot be shared across pods.
+
+**Logs**
+
+Kubernetes stores the logs of each container running in `/var/logs`.  These logs are ephemeral and are monitored by Kubelet to make sure that they do not grow too large while the pods are running.
+
+**Container runtime**
+
+The container runtime has two different areas of storage for containers and images. Images are usually denoted as the read-only layer, as they are not modified when containers are running. There is a thin layer on top of containers that provides ephemeral storage for containers if the container is writing to the filesystem. This is referred to as the writable layer.
+
+## Container Runtime Configurations
+
+### CRI-O
+
+CRI-O uses a storage configuration file to control how the container runtime stores persistent and temporary data.
+CRI-O utilizes the [storage library](https://github.com/containers/storage).  
+Some Linux distributions have a manual entry for storage (`man 5 containers-storage.conf`).
+The main configuration for storage is located in `/etc/containers/storage.conf` and one can control the location for temporary data and the root directory.  
+The root directory is where CRI-O stores the persistent data.
+
+```toml
+[storage]
+# Default Storage Driver
+driver = "overlay"
+# Temporary storage location
+runroot = "/var/run/containers/storage"
+# Primary Read/Write location of container storage 
+graphroot = "/var/lib/containers/storage"
+```
+
+- graphroot
+  - Persistent data stored from the container runtime
+  - If SELinux is enabled, this must match the `/var/lib/containers/storage`
+- runroot
+  - Temporary read/write access for container.
+  - Recommended to have this on a temporary filesystem.
+
+A quick way to relabel your graphroot directory to match `/var/lib/containers/storage`.
+
+```bash
+semanage fcontext -a -e /var/lib/containers/storage /YOUR_STORAGE_PATH
+restorecon -R -v /YOUR_STORAGE_PATH
+```
+
+### Containerd
+
+containerd uses a toml configuration file to control where persistent and ephemeral data is stored.
+The default path for the config file is located at `/etc/containerd/config.toml`.
+
```

Comment:
@samuelkarp What is your suggestion for this? Is this for my information? Or do you want to add this to the containerd section? I'm happy to give you access to my PR if you want to add more details to the containerd section.

Reply:
Is it possible to add some suggestion to add `volatile` mount for overlayfs filesystem?
I think it's very useful feature because kubernetes doesn't allow to restart exited container.
For container's rootfs, we don't need sync-fs during umount rootfs. It can reduce IO impact for running containers or controller, like kubelet and containerd/cri-o.

Reply:
I was not trying to be exhaustive with the container runtime setup for storage. I don't think that should be the goal of the post. I wanted to mention how containerd can have a separate container runtime filesystem from node filesystem.

Reply:
it's fine. The goal to separate image filesytem is to reduce IO impact, at least to me.